pages/_app/index.jsx

```diff
@@ -110,17 +110,22 @@ class App extends NextApp {
         })
 
         // Subscribe store to router after initialization
-        router.events.on('routeChangeStart', this.handleRouteChange)
+        // We use routeChangeComplete otherwise router.query contains previous
+        // values
+        router.events.on('routeChangeComplete', this.handleRouteChange)
 
-        // Trigger App render
+        store.changeDataProvider(dataProviderId)
         this.setState({ isAuthorized: true })
       })
-      .catch((error) => {
+      .catch(async (error) => {
         if (error instanceof AuthorizationError) {
           this.setState({ isAuthorized: false })
           this.redirectToLogin({
             reason: '',
           })
+        } else if (error instanceof AccessError) {
+          await router.replace('/data-providers')
+          this.setState({ isAuthorized: true })
```

Comment:
What do you wait for the state changing? I will cause extra rendering and may cause extra redirecting. I am not sure this is what you want.

I would do it like this:
```js
          router.replace('/data-providers')
          this.setState({ isAuthorized: true })
```

or like this:
```js
          this.setState({ isAuthorized: true }, () => router.replace('/data-providers'))
```

What you did is like

```js
          router.replace('/data-providers').then(() => this.setState({ isAuthorized: true })
```

what is also fine but I am wondering why the order should be like this.

Reply:
I should have probably write a comment there it's tricky. Sorry about that. 

If the state is updated before route changes nextjs may render the old component. Since we don't have any data for that provider it will fail with error because Works, DepositDates are not initialized.

Reply:
It's very unlikely because the production build is really fast. Nonetheless it should be covered.